### PR TITLE
livecheck: raise error if no watchlist

### PR DESCRIPTION
**Pull request by hyuraku**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?
- [x] Have you successfully run `brew man` locally and committed any changes?

-----
closes https://github.com/Homebrew/brew/issues/10052
raise an error when `brew livecheck` has no args and set no `ENV['HOMEBREW_LIVECHECK_WATCHLIST'] `or `~/.brew_livecheck_watchlist`

### CODE REVIEW

**Comment by samford** [comment]
This issue was introduced in #8578 (adding cask support). Before that point, when a user ran `brew livecheck` without `HOMEBREW_LIVECHECK_WATCHLIST` set or a `~/.brew_livecheck_watchlist` file existing, they would see the following:

```
$ brew livecheck
Usage: brew livecheck [formulae]

Check for newer versions of formulae from upstream.

If no formula argument is passed, the list of formulae to check is taken from
HOMEBREW_LIVECHECK_WATCHLIST or ~/.brew_livecheck_watchlist.

      --full-name                  Print formulae with fully-qualified names.
      --tap                        Check formulae within the given tap,
                                   specified as user/repo.
      --all                        Check all available formulae.
      --installed                  Check formulae that are currently
                                   installed.
      --newer-only                 Show the latest version only if it's newer
                                   than the formula.
      --json                       Output information in JSON format.
  -q, --quiet                      Suppress warnings, don't print a progress
                                   bar for JSON output.
  -d, --debug                      Display any debugging information.
  -v, --verbose                    Make some output more verbose.
  -h, --help                       Show this message.

Error: Invalid usage: No formulae to check.
```

I pushed a commit that reinstates this behavior by using the safe navigation operator before `sort_by` (i.e., `end&.sort_by`). [I separated the `sort_by` call from the if/else block since `brew style` wanted the `if`/`elsif`/`else` to all be in line with the addition of the safe navigation operator and that would have been a mess. It's possible to move the `if` onto the next line and indent everything two spaces but having the `sort_by` block separate helps to improve clarity anyway.]

That said, the question is whether the existing `Error: Invalid usage: No formulae to check.` message is sufficient or would we benefit from the more specific `UsageError` in this PR. On the one hand, the usage banner already contains information about the watchlist. On the other hand, users may overlook that part of the usage banner and making it clear the error is related to a missing watchlist file may be helpful to new users.

Maybe we could split the difference and make the watchlist `UsageError` a bit more brief and rely on the usage banner for the specific information. Do we think something like `A watchlist file is required when no arguments are given.` would be sufficient?

**Review by nandahkrishna — COMMENTED** [review]
> `A watchlist file is required when no arguments are given.`

I prefer this because it's possible that the user would miss the information present in the usage banner, and `No formulae to check` doesn't seem specific enough.

File: Library/Homebrew/dev-cmd/livecheck.rb [hunk 1]
```diff
@@ -60,44 +60,48 @@ def livecheck
...
+              (args.cask? && !formula_or_cask.is_a?(Cask::Cask))
+          end
+        rescue Errno::ENOENT => e
+          onoe e
         end
-      rescue Errno::ENOENT => e
-        onoe e
+      else
+        raise UsageError, "ENV['HOMEBREW_LIVECHECK_WATCHLIST'] or ~/.brew_livecheck_watchlist is required " \
+                          "if no formula or cask argument is passed"
```
Comment (samford) [inline Library/Homebrew/dev-cmd/livecheck.rb:101]:
```suggestion
        raise UsageError, "A watchlist file is required when no arguments are given."
```

**Review by samford — CHANGES_REQUESTED** [review]
I agree that there's value in having a better error than `No formulae or casks to check.` in this context. I imagine we're in agreement on this point, so it's just a matter of figuring out the best error message.

I think the suggestion below is a happy medium and users can simply refer to the usage banner if they're not familiar with watchlists.

**Review by samford — APPROVED** [review]

**Review by nandahkrishna — APPROVED** [review]

**Comment by samford** [comment]
I'm going to go ahead and merge this since it's fixing a bug. Thanks, @hyuraku!

**Pull request merged**